gateway/src/apicast/policy/headers.lua

```diff
@@ -0,0 +1,89 @@
+--- Headers policy
+-- This policy allows to include custom headers that will be sent to the
+-- upstream as well as delete headers included in the request that should not
+-- be sent to the upstream.
+-- Similarly, this policy also allows to add and delete headers of the
+-- response.
+
+local ipairs = ipairs
+local pairs = pairs
+
+local policy = require('apicast.policy')
+local _M = policy.new('Headers policy')
+
+local new = _M.new
+
+local function add_req_headers(headers)
+  for header, value in pairs(headers) do
+    ngx.req.set_header(header, value)
+  end
+end
+
+local function delete_req_headers(headers)
+  for _, header in ipairs(headers) do
+    ngx.req.clear_header(header)
+  end
+end
+
+local function add_resp_headers(headers)
+  for header, value in pairs(headers) do
```

Comment:
`pairs` can't be JITed. So we should either use Luafun or make it a numerically indexed table.
I know this is pain, but that is a limitation of LuaJIT.

Reply:
I'd rather evaluate Luafun later.
Do you see any other option apart from having two separate arrays in the config? Something like this:
```
          {
            "name": "apicast.policy.headers",
            "configuration":
              {
                "response":
                  {
                    "headers_to_add": { "Header1", "Header2" }
                    "values_of_headers_to_add": { "value1", "value2" }
                  }
              }
          }
```

Reply:
I was thinking about the configuration a bit and it really depends how the UI will look like.
But something like this seemed flexible:


```json
 {
            "name": "apicast.policy.headers",
            "configuration":
              {
                "response":
                  [
                    { "name": "Header1", "op": "set", "value": "foobar" },
                    { "name": "Header1", "op": "add", "value": "foobar1" },
                    { "name": "Header2", "op": "remove" },
                  ]
              }
          }
```

Reply:
That could work :+1: 
Also, I see "add" there. What would that do? Add only if the header has not been set already?

Reply:
@davidor `add` is the array operation (in my mind). Because all headers that have `,` are technically arrays and can be sent as individual headers:

```
Set-Cookie: foo=bar, one=two
```

Is the same as:

```
Set-Cookie: foo=bar
Set-Cookie: one=two
```

Reply:
I see. I didn't get that you were refering to 'add' as the array operation :+1:

Reply:
I'm thinking about the difference between `add` and `set` in you're example.

Do you think that `add` should fail if there's no a `set` op for the same header before or done by Apicast? If that's not the case, having just `add` or `set` would be enough.

I don't see a clear use case where we'd like to add a value separated by ',' in a header only if that header has been already defined.

Reply:
I think `add` should not care if the value exists. It just should add to it. Its purpose is adding header values without the risk of overriding them. 

Now I wonder if `remove` can be implemented in terms of `set` (empty value). Not sure if nginx is going to send the empty header or not.

Reply:
How would 'set' work in the case of repeated headers like your cookie example?
Blows away the entire array of values and sets a new value (or an array of values if that is allowed?).

I assume 'remove' removes all values if the header was an array?

Is it possible to 'add' an array of values?

Reply:
@andrewdavidmackenzie 
If I understood correctly:
1) Yes. I think that to be coherent with @mikz definition, we should delete the exiting array of values and initialize it with the value passed to the set op.
2) Yes.
3) That's something we need to define.

Reply:
Ok, I'll try to spec it out in more detail.

# set

Ability to override any value (including non existing) to a specific value. Removes all previous values and sets the one desired.

```json
{ "op": "set", "name": "X-Header", "value": "foo" }
```

Examples:


```
GET /
X-Header: bar, daz
```

will be transformed to:

```
GET /
X-Header: foo
```

Request:
```
GET /
X-Foo: sometihing
```

will be transformed to:
```
GET /
X-Foo: something
X-Header: foo
```

# remove

Can be implemented as `set` with value of empty string. Nginx will not send out empty headers.


```json
{ "op": "set", "name": "X-Header", "value": "" }
```

# add


```json
{ "op": "add", "name": "X-Header", "value": "foo" }
```

Will append `, value` to the existing value of the header or set a new `value` if there is no previous value. Duplicates allowed.

request

```
GET /
```

transformed to

```
GET / 
X-Header: foo
```

request

```
GET /
X-Header: foo
```

transformed to

```
GET
X-Header: foo, foo
```

Reply:
Perfect :ok_hand: 

There's only one question left. @andrewdavidmackenzie wrote it above. Should we allow:
```
{ "op": "add", "name": "X-Header", "value": "abc, def" }
```

To transform the request
```
GET /
X-Header: foo
```

Into:
```
GET /
X-Header: foo, abc, def
```

?

Reply:
@davidor I think that is fine and within the specs. There is no escaping of `,`. Less rules - better :)